### PR TITLE
Add `replace_null` API with `replace_policy` parameter, `fixed_width` column support

### LINKED ISSUE #1361: [FEA] Add forward and backward fill options to DataFrame.fillna
It would be nice to have forward fill and backward fill options within the DataFrame.fillna method to use in time series applications. The "method" argument is already implemented in the API currently.

```python
import cudf
df = cudf.DataFrame({'x': [1, 2, None, 4]})
df.fillna(method='ffill') # returns df with None filled as 2
df.fillna(method='bfill') # returns df with None filled as 4
```

This issue is somewhat of a continuation of #1106 .

### PR DESCRIPTION
**Pull request by isVoid**
Part 1 for issue #1361 

- Adds `PRECEDING` and `FOLLOWING` options to `replace_nulls` in `libcudf`. This PR provides support for `fixed_width_type` type columns.
- Adds Cython binding

### CODE REVIEW

**Comment by GPUtester** [comment]
Please update the changelog in order to start CI tests.

View the gpuCI docs [here](https://docs.rapids.ai/gpuci#open-pr-or-update-pr).

File: cpp/include/cudf/replace.hpp [hunk 1]
```diff
@@ -26,6 +26,14 @@ namespace cudf {
  * @file
  */
 
+/**
+ * @brief Policy to fill null/nan values inside
+ *
+ * `FORWARD_FILL` means filling the null
+ * `BACKWARD_FILL` means
+ */
+enum class fillna_policy : bool { FORWARD_FILL, BACKWARD_FILL };
```
Comment (jrhemstad) [inline cpp/include/cudf/replace.hpp:35]: We should strive to have more descriptive names in libcudf than just reusing Pandas terminology. 
```suggestion
enum class replace_policy : bool {PRECEDING, FOLLOWING};
```

File: cpp/src/replace/nulls.cu [hunk 2]
```diff
@@ -352,6 +355,76 @@ std::unique_ptr<cudf::column> replace_nulls_scalar_kernel_forwarder::operator()<
...
+                                           rmm::cuda_stream_view stream,
+                                           rmm::mr::device_memory_resource* mr)
+  {
+    // Construct input data-validity zip iterator
+    using Type     = cudf::device_storage_type_t<col_type>;
+    auto device_in = cudf::column_device_view::create(input);
+
+    auto in_validity_iterator = cudf::detail::make_validity_iterator(*device_in);
+    auto incol_validity_zip_iterator =
+      thrust::make_zip_iterator(thrust::make_tuple(input.begin<Type>(), in_validity_iterator));
```
Comment (jrhemstad) [inline cpp/src/replace/nulls.cu:385]: Use `make_pair_iterator`.

File: cpp/src/replace/nulls.cu [hunk 3]
```diff
@@ -352,6 +356,75 @@ std::unique_ptr<cudf::column> replace_nulls_scalar_kernel_forwarder::operator()<
...
+                                           rmm::cuda_stream_view stream,
+                                           rmm::mr::device_memory_resource* mr)
+  {
+    using Type     = cudf::device_storage_type_t<col_type>;
+    auto device_in = cudf::column_device_view::create(input);
+    auto index     = thrust::make_counting_iterator<cudf::size_type>(0);
+    auto valid_it  = cudf::detail::make_validity_iterator(*device_in);
+    auto in_begin  = thrust::make_zip_iterator(thrust::make_tuple(index, valid_it));
+
+    rmm::device_vector<cudf::size_type> gather_map(input.size());
```
Comment (jrhemstad) [inline cpp/src/replace/nulls.cu:387]:
```suggestion
    rmm::device_uvector<cudf::size_type> gather_map(input.size());
```

File: cpp/src/replace/nulls.cu [hunk 4]
```diff
@@ -352,6 +356,75 @@ std::unique_ptr<cudf::column> replace_nulls_scalar_kernel_forwarder::operator()<
   return cudf::dictionary::detail::replace_nulls(dict_input, replacement, stream, mr);
 }
 
+struct replace_nulls_replace_policy_functor {
```
Comment (jrhemstad) [inline cpp/src/replace/nulls.cu:359]: This name is awfully verbose.

File: cpp/src/replace/nulls.cu (hunk 3 above)
Comment (jrhemstad) [reply cpp/src/replace/nulls.cu:387]: I think you want to initialize this gather_map with some sort of sentinel value that is out of bounds so elements in the output that should be null get nullified in the gather.

File: cpp/src/replace/nulls.cu (hunk 4 above)
Comment (isVoid) [reply cpp/src/replace/nulls.cu:359]: `replace_policy_functor`?

**Comment by isVoid** [comment]
FYI I will separate python API changes to a follow-up PR. It makes more sense to make such changes once the Cython bindings are stable.

File: cpp/src/replace/nulls.cu (hunk 4 above)
Comment (isVoid) [reply cpp/src/replace/nulls.cu:359]: Looking at other class/test names, there is certainly a better naming for them as well.

File: cpp/src/replace/nulls.cu [hunk 5]
```diff
@@ -352,6 +356,75 @@ std::unique_ptr<cudf::column> replace_nulls_scalar_kernel_forwarder::operator()<
...
+                             gm_rbegin,
+                             func);
+    }
+
+    cudf::column_view gm_col(cudf::data_type{cudf::type_id::INT32},
+                             gather_map.size(),
+                             thrust::raw_pointer_cast(gather_map.data()));
+
+    auto output =
+      cudf::gather(cudf::table_view({input}), gm_col, cudf::out_of_bounds_policy::DONT_CHECK, mr);
```
Comment (jrhemstad) [inline cpp/src/replace/nulls.cu:413]: Just use the `detail::gather` that works on iterators instead of constructing a `column_view` from the gather map.

**Review by jrhemstad — CHANGES_REQUESTED** [review]
Does this work in the case where an output element should end up null due to all the preceding (or following) elements being null? I think you'd need to initialize the gather map to some out of bounds sentinel value and pass `NULLIFY` to gather to ensure those elements end up nullified.

**Comment by isVoid** [comment]
> Does this work in the case where an output element should end up null due to all the preceding (or following) elements being null? I think you'd need to initialize the gather map to some out of bounds sentinel value and pass `NULLIFY` to gather to ensure those elements end up nullified.

In fact, current solution will trivially work. Example `PRECEDING` fill with leading null:
```text
Input: NULL NULL -10 NULL -30
Gather map after inclusive scan: 0 0 2 2 4
Result: NULL NULL -10 -10 -30
```
Thrust's `inclusive_scan` copies the first element and if it is `NULL`, it gets propagated to all `NULL`s that follows.

Test case for `PRECEDING` fill with leading nulls and `FOLLOWING` fill with trailing nulls are shown here:
https://github.com/rapidsai/cudf/blob/c3e6b3daefb5283238ff0855a06a95f9363fb668/cpp/tests/replace/replace_nulls_tests.cpp#L351
https://github.com/rapidsai/cudf/blob/c3e6b3daefb5283238ff0855a06a95f9363fb668/cpp/tests/replace/replace_nulls_tests.cpp#L368

**Comment by jrhemstad** [comment]
> > Does this work in the case where an output element should end up null due to all the preceding (or following) elements being null? I think you'd need to initialize the gather map to some out of bounds sentinel value and pass `NULLIFY` to gather to ensure those elements end up nullified.
> 
> In fact, current solution will trivially work. Example `PRECEDING` fill with leading null:
> 
> ```
> Input: NULL NULL -10 NULL -30
> Gather map after inclusive scan: 0 0 2 2 4
> Result: NULL NULL -10 -10 -30
> ```
> 
> Thrust's `inclusive_scan` copies the first element and if it is `NULL`, it gets propagated to all `NULL`s that follows.
> 
> Test case for `PRECEDING` fill with leading nulls and `FOLLOWING` fill with trailing nulls are shown here:
> https://github.com/rapidsai/cudf/blob/c3e6b3daefb5283238ff0855a06a95f9363fb668/cpp/tests/replace/replace_nulls_tests.cpp#L351
> 
> 
> https://github.com/rapidsai/cudf/blob/c3e6b3daefb5283238ff0855a06a95f9363fb668/cpp/tests/replace/replace_nulls_tests.cpp#L368
> 
> @jrhemstad

Oh! Good point! If all the preceding values were null, the gather map will contain an index that points at a null element and so the gather will just gather the null element. Makes sense.

**Review by shwina — APPROVED** [review]
Cython LGTM!

**Review by jrhemstad — APPROVED** [review]

File: cpp/include/cudf/replace.hpp [hunk 6]
```diff
@@ -26,6 +26,14 @@ namespace cudf {
  * @file
  */
 
+/**
+ * @brief Policy to specify the relative position that replacement value is to the null row
```
Comment (harrism) [inline cpp/include/cudf/replace.hpp:30]: Clearer.

```suggestion
 * @brief Policy to specify the position of replacement values relative to null rows
```

File: cpp/include/cudf/replace.hpp [hunk 7]
```diff
@@ -61,6 +69,23 @@ std::unique_ptr<column> replace_nulls(
...
   rmm::mr::device_memory_resource* mr = rmm::mr::get_current_device_resource());
 
+/**
+ * @brief Replaces all null values in a column with the first non-null value that precedes/follows.
+ *
+ * If `input[i]` is NULL, then `output[i]` will contain the first non-null value that precedes or
+ * follows the null value, based on `replace_policy`.
+ *
+ * @param[in] input A column whose null values will be replaced.
+ * @param[in] replace_policy Specify the position that replacement value is to the null value.
```
Comment (harrism) [inline cpp/include/cudf/replace.hpp:79]:
```suggestion
 * @param[in] replace_policy Specify the position of replacement values relative to null values.
```

File: cpp/src/replace/nulls.cu [hunk 8]
```diff
@@ -352,6 +357,67 @@ std::unique_ptr<cudf::column> replace_nulls_scalar_kernel_forwarder::operator()<
   return cudf::dictionary::detail::replace_nulls(dict_input, replacement, stream, mr);
 }
 
+struct replace_policy_functor {
```
Comment (harrism) [inline cpp/src/replace/nulls.cu:360]: Doc? Not a public doc, but explain what this is for.

**Review by harrism — APPROVED** [review]

**Review by harrism — CHANGES_REQUESTED** [review]
Clicked the wrong button. Just some doc requests.

File: cpp/include/cudf/replace.hpp (hunk 6 above)
Comment (isVoid) [reply cpp/include/cudf/replace.hpp:30]: Lol, I literally rephrased this 2-3 times but still couldn't get it right.

**Review by harrism — APPROVED** [review]

**Pull request merged**